.github/workflows/publish.yml

```diff
@@ -1,21 +1,52 @@
-name: publish
+name: Build and Publish Docker Image
 
 on:
   push:
-    branches: [master, develop]
+    branches:
+        - "master"
+        - "develop"
+    tags:
+      - "v*.*.*"
 
 jobs:
-  publish-docker-image:
+  docker:
     runs-on: ubuntu-latest
     steps:
-      - name: Checkout the repository
+      -
+        name: Checkout
         uses: actions/checkout@v3
-
-      - name: Build and publish the Docker image
-        run: |
-          echo $GITHUB_TOKEN | docker login ghcr.io -u center-for-threat-informed-defense --password-stdin
-          docker build . --tag ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:$IMAGE_TAG
-          docker push ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api:$IMAGE_TAG
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          IMAGE_TAG: "${{ github.ref == 'refs/head/master' && 'latest' || 'develop' }}"
+      -
+        name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: |
+            ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api
+          tags: |
+            # set latest tag for master branch
+            type=raw,value=latest,enable=${{ github.ref == format('refs/heads/{0}', 'master') }}
+            # set develop tag for develop branch
+            type=ref,event=branch,enable=${{ github.ref == format('refs/heads/{0}', 'develop') }}
+            # set semver tag (vX.Y.Z) on git tag event
+            type=semver,pattern=v{{major}}.{{minor}}.{{patch}}
+            # set git short commit as Docker tag (e.g., sha-ad132f5)
+            type=sha
```

Comment:
This action sets the Docker image name:
- `ghcr.io/center-for-threat-informed-defense/attack-workbench-rest-api`

and dynamically sets Docker tags based on the type of git event.

The `Build and push` job at the end of this workflow will reference the Docker image name and any Docker tags produced via `steps.meta`.

Here is a breakdown of how we dynamically generate the Docker tags:

### latest
```
type=raw,value=latest,enable=${{ github.ref == format('refs/heads/{0}', 'master') }}
```
- [Here](https://github.com/docker/metadata-action#typeraw) is the documentation on the `raw` tag type. In short, it lets you explicitly define string literals as Docker tags.
- This one sets the `latest` tag, but it will only take effect when the workflow is executing on the `master` branch, as dictated by `enable=${{ github.ref == format('refs/heads/{0}', 'master') }}`.

### develop
```
type=ref,event=branch,enable=${{ github.ref == format('refs/heads/{0}', 'develop') }}
```
- This one works identically as the one above, except that it will set the **branch name** as the Docker tag.
- Again, we use the `enable` property to ensure that this tag only activates when the workflow is running against the `develop` branch.

### semver
```
type=semver,pattern=v{{major}}.{{minor}}.{{patch}}
```
- [Here](https://github.com/docker/metadata-action#typesemver) is the documentation for the `semver` tag type.
- It runs on a push tag event and requires a valid semver string as the git tag.
- e.g., Will run on `git tag v1.0.0` but will NOT run on `git tag foobarv1`
- This action will parse the git tag string and express several variables that we can reference programmatically to set the desired Docker tag. 
- Here, we use the `major`, `minor`, and `patch` variables to set the Docker tag using string interpolation.
- Here's how it would work in action:
    - Developer pushes git tag `v1.2.3` to repository `foo/bar`.
    - The workflow parses `v.1.2.3` and sets `major=1`, `minor=2`, and `patch=3`. These variable are accessible during the GitHub Actions workflow.
    - We use the aforementioned variables to set a Docker tag using `pattern=v{{major}){{minor}}{{patch}}` because we want to ensure that the `v` prefix is included.
    - Docker image `foo/bar:v1.2.3` would be produced.
- There are several ways to accomplish the same thing. For example, we could use the `raw` variable because the action will only execute on a valid semver tag, and the `raw` pattern refers to the git tag string itself, which we already know to be semver compliant. This approach is just more explicit.

```
type=sha
```
- [Here](https://github.com/docker/metadata-action#typesha) is the documentation for the `type=sha` flavor.
- It just sets the git short commit hash value as a Docker tag. 
- This can be useful for debugging purposes if we're trying to match up a given Docker image to the respective git commit so we can identify the source code that went into building that particular image.